Add kops documentation (#1055)
aws/karpenter: website/content/en/docs/getting-started-with-kops/_index.md

---
title: "Getting Started with Karpenter using kOps"

> linkTitle: "Getting Started with kOps"

Reviewer: Can we make the title the same as the linkTitle?

weight: 10
---

In this example, the cluster will be running on Amazon Web Services (AWS) managed by [kOps](https://kops.sigs.k8s.io/).

> Karpenter is designed to be cloud provider agnostic, but currently only supports AWS. Contributions are welcomed

Reviewer 2: Not sure about `welcome` vs `welcomed`, but at least a period:
```suggestion
Karpenter is designed to be cloud provider agnostic, but currently only supports AWS. Contributions are welcomed.
```

> Karpenter is supported on kOps as of 1.24.0-alpha.2, but sits behind a feature flag as the interface between kOps and Karpenter is

Reviewer: I'm not sure about the long term viability of wrapping Karpenter's API with an instance group and the conflicting concerns between whether or not Karpenter or Kops owns the LT. Can we be ultra crisp about this and let customers know that they should only take a dependency on this alpha kops integration if they're prepared for significant incompatible changes?

Author: Emphasized it some more

> still work in progress. This guide is intended for users that wants to test Karpenter on kOps and provide feedback to Karpenter and kOps developers.

Reviewer 2: (Google favors `a` 8:6, which admittedly isn't a huge difference, but...)
```suggestion
still a work in progress. This guide is intended for users who want to test Karpenter on kOps and provide feedback to Karpenter and kOps developers.
```

Read more about how Karpenter works on kOps and the current limitations in the [kOPs Karpenter documentation](https://kops.sigs.k8s.io/operations/karpenter/).

This guide should take less than 1 hour to complete, and cost less than $0.25.
Follow the clean-up instructions to reduce any charges.

This guide assumes you already have a kOps state store and a hosted zone. If you do not have one,
run through the [kOps getting started on AWS documentation](https://kops.sigs.k8s.io/getting_started/aws/) up until "Creating your first cluster".

## Install

Karpenter is installed in clusters as a managed addon. kOps will automatically create 

> and manage the necessary the IAM roles and policies Karpenter needs.

Reviewer 2: ```suggestion
and manage the necessary IAM roles and policies Karpenter needs.
```

### Required Utilities

Install these tools before proceeding:

1. `kubectl` - [the Kubernetes CLI](https://kubernetes.io/docs/tasks/tools/install-kubectl-linux/)
2. `kops` - [kubectl, but for clusters](https://kops.sigs.k8s.io/getting_started/install/) v1.24.0 or later

### Environment Variables

After setting up the tools, set the following environment variables used by kOps.

```bash
export KOPS_FEATURE_FLAGS=Karpenter
export CLUSTER_NAME=${USER}-karpenter-demo.example.com
export ZONES=us-west-2a
export KOPS_STATE_STORE=s3://prefix-example-com-state-store
export KOPS_OIDC_STORE=s3://prefix-example-com-oidc-store/discovery
```

### Create a Cluster

> kOps installs Karpenter on the control plane. Once the control plane is running, Karpenter will provision the
> the worker nodes needed for non-Control Plane Deployments such as CoreDNS and CSI drivers.

Reviewer 2: ```suggestion
kOps installs Karpenter on the control plane. Once the control plane is running, Karpenter will provision
the worker nodes needed for non-Control Plane Deployments such as CoreDNS and CSI drivers.
```

The following command will launch a cluster with Karpenter-managed worker nodes:

```bash
kops create cluster \
    --zones=$ZONES \
    --discovery-store=${KOPS_OIDC_STORE} \
    --instance-manager=karpenter \
    --networking=amazonvpc \
    ${CLUSTER_NAME} \
    --yes
```

> Note: we are using AWS VPC CNI for networking as Karpenter's binpacking logic assumes ENI-based networking.

Reviewer 2: Personally I find this a bit too buzzwordy, but if it's reasonable to expect a reader to know both CNI and ENI, then I guess it's ok. If it's likely a reader will know only one, or potentially neither, then I'd suggest links for both.

Author: I think either they are familiar with the kOps networking options, have opinions and want an explanation for why this CNI is used, in which case I do expect them to be familiar with these terms. Or they are not familiar with them, and doesn't have any opinions.

Ideally, this ENI/max-pods limitation is documented elsewhere on this site, and I could just link to an explanation there.

> ### Create Grafana dashboards (optional)

Reviewer: We're planning on making a separate getting started guide for grafana. I'd leave this out of here for now.

The Karpenter repo contains multiple [importable dashboards](https://github.com/aws/karpenter/tree/main/grafana-dashboards) for an existing Grafana instance. See the Grafana documentation for [instructions](https://grafana.com/docs/grafana/latest/dashboards/export-import/#import-dashboard) to import a dashboard.

> #### Deploy a temporary Prometheus and Grafana stack (optional)

Reviewer: likewise

> The following commands will deploy a Prometheus and Grafana stack that is suitable for this guide but does not include persistent storage or other configurations that would be necessary for monitoring a production deployment of Karpenter.

Reviewer 2: ```suggestion
The following commands will deploy a Prometheus and Grafana stack that is suitable for this guide but does not include persistent storage or other configurations that would be necessary for monitoring a production deployment of Karpenter:
```

```sh
helm repo add grafana-charts https://grafana.github.io/helm-charts
helm repo add prometheus-community https://prometheus-community.github.io/helm-charts
helm repo update

kubectl create namespace monitoring

curl -fsSL https://karpenter.sh/docs/getting-started/prometheus-values.yaml | tee prometheus-values.yaml
helm install --namespace monitoring prometheus prometheus-community/prometheus --values prometheus-values.yaml

curl -fsSL https://karpenter.sh/docs/getting-started/grafana-values.yaml | tee grafana-values.yaml
helm install --namespace monitoring grafana grafana-charts/grafana --values grafana-values.yaml
```

> The Grafana instance may be accessed using port forwarding.

Reviewer 2: ```suggestion
The Grafana instance may be accessed using port forwarding:
```

```sh
kubectl port-forward --namespace monitoring svc/grafana 3000:80
```

> The new stack has only one user, `admin`, and the password is stored in a secret. The following command will retrieve the password.

Reviewer 2: ```suggestion
The new stack has only one user, `admin`, and the password is stored in a secret. The following command will retrieve the password:
```

```sh
kubectl get secret --namespace monitoring grafana -o jsonpath="{.data.admin-password}" | base64 --decode
```

### Provisioner

A single Karpenter provisioner is capable of handling many different pod
shapes. Karpenter makes scheduling and provisioning decisions based on pod
attributes such as labels and affinity. In other words, Karpenter eliminates
the need to manage many different InstanceGroups.

kOps manage provisioners through InstanceGroups. Your cluster will already have
one Provisioner that will contain a suitable set of instance types for Karpenter to
choose from.

Managing Provisioner resources directly is possible, but not straight-forward. Read

> more about managing provisioners in the [kOPs Karpenter documentation](https://kops.sigs.k8s.io/operations/karpenter/)

Reviewer 2: ```suggestion
more about managing provisioners in the [kOPs Karpenter documentation](https://kops.sigs.k8s.io/operations/karpenter/).
```

other kOps addons. Create additional pods using a Deployment, and watch Karpenter provision nodes in response.

### Automatic Node Provisioning

This deployment uses the [pause image](https://www.ianlewis.org/en/almighty-pause-container) and starts with zero replicas.

```bash
cat <<EOF | kubectl apply -f -
apiVersion: apps/v1
kind: Deployment
metadata:
  name: inflate
spec:
  replicas: 0
  selector:
    matchLabels:
      app: inflate
  template:
    metadata:
      labels:
        app: inflate
    spec:
      terminationGracePeriodSeconds: 0
      containers:
        - name: inflate
          image: public.ecr.aws/eks-distro/kubernetes/pause:3.2
          resources:
            requests:
              cpu: 1
EOF
kubectl scale deployment inflate --replicas 5
kubectl logs -f -n kube-system -l karpenter=controller
```

### Automatic Node Termination

Now, delete the deployment. After 30 seconds,
Karpenter should terminate the now empty nodes.

```bash
kubectl delete deployment inflate
kubectl logs -f -n karpenter -l karpenter=controller 
```

### Manual Node Termination

If you delete a node with kubectl, Karpenter will gracefully cordon, drain,

> and shutdown the corresponding instance. Under the hood, Karpenter adds a

Reviewer 2: ```suggestion
and shut down the corresponding instance. Under the hood, Karpenter adds a
```

> finalizer to the node object, which blocks deletion until all pods are

Reviewer 2: ```suggestion
finalizer to the node object which blocks deletion until all pods are
```

drained and the instance is terminated. Keep in mind, this only works for
nodes provisioned by Karpenter.

```bash
kubectl delete node $NODE_NAME
```

This is similar to `kops delete instance $NODE_NAME` except for that kOps will not respect
`karpenter.sh/do-not-evict`.

## Upgrading a Cluster

kOps is aware of nodes managed by Karpenter and will handle [rolling upgrades](https://kops.sigs.k8s.io/operations/rolling-update/) of those nodes the same way as any other node:

```
kops upgrade cluster --yes
kops update cluster --yes
kops rolling-update cluster --yes
```

Karpenter-managed InstanceGroups supports setting `maxUnavailable`, but since Karpenter instances do not run in an Auto Scaling Group, setting `maxSurge` will not have any effect.

> During rolling updates, `karpenter.sh/do-not-evict` is not respected.

Reviewer: This makes me very sad. Why can't kops use `k delete node` to roll the instance group?

Author: This is entirely a design decision. Now, I landed on consistent behavior within a kOps cluster rather than consistency across Karpenter clusters.
There shouldn't be a problem to run `delete node` instead of `drain node` on Karpenter InstanceGroups. The question is what the user expects and what makes sense. We are in an active user-initiated upgrade process. Typically performed in maintenance windows. We frequently have issues from people who want us to ignore PDBs, or some sort of timeout on PDB violations, so usually it seems like the behavior users want is to get upgrade done as fast and graceful as possible rather than being endlessly blocked by some pod some user put an annotation on (not necessarily knowing the consequence of that).

Anyways, the documentation is status quo and can be changed if users prefer/expect different behavior. Note that we do not respect CAS' safe to evict-flag either.

## Cleanup

> To avoid additional charges, remove the demo infrastructure from your AWS account.

Reviewer 2: ```suggestion
To avoid additional charges, remove the demo infrastructure from your AWS account:
```
